design/metadata_kv/index.md

```diff
@@ -0,0 +1,187 @@
+
+Concurrency is more of an issue here because of how a commit works: when a commit starts, it scans the currently staged changes, applies them to the current commit pointed to by the branch, updating the branch reference and removing the staged changes it applied.
+
+Getting this right means we have to take care of the following:
+
+1. Ensure all staged changes that finished successfully before the commit started are applied as part of the commit (causality)
+1. Ensure acknowledged writes end up either in a resulting commit, or staged to be committed (no lost writes)
+
+To do this, we will employ 2 mechanisms:
+
+1. [Optimistic Concurrency Control](https://en.wikipedia.org/wiki/Optimistic_concurrency_control) on the branch pointer using `SetIf()`
+1. Reliance on write [idempotency](https://en.wikipedia.org/wiki/Idempotence) provided by Graveler (i.e., Writing the same exact entry, with the same identity - will not appear as a change)
+
+This is what the proposed implementation will look like:
+
+#### Committer flow
+
+We add an additional field to each `Branch` object: In addition to the existing `staging_token`, we add an array of strings named `sealed_tokens`.
+
+1. get branch, find current `staging_token`
+1. use `SetIf()` to update the branch (if not modified by another process): push existing `staging_token` into `sealed_tokens`, set new uuid as `staging_token`.
+1. take the list of sealed tokens, and using the [`CombinedIterator()`](https://github.com/treeverse/lakeFS/blob/master/pkg/graveler/combined_iterator.go#L11), turn them into a single iterator to be applied on top of the existing commit
+1. Once the commit has been persisted (metaranges and ranges stored in object store, commit itself stored to KV using `Set()`), perform another `SetIf()` that updates the branch again: replacing its commit ID with the new value, and clearing `sealed_tokens`, as these have materialized into the new commit.
+1. If `SetIf()` fails, this means another commit is happenning/has happened on the same branch. Can either retry or let the caller know.
+
+An important property here is delegating safety vs liveness to a single optimistic `SetIf()` operation: if a commit fails somewhere along the process, a subsequent commit would simply pick up where the failed one left off,
+adding the current staging_token into the set of changes to apply. In an environment where there aren't many concurrent commits on the same branch, and that commits mostly succeed - the size of `sealed_tokens` would be relatively small. As an optimization, compaction strategies could be added to merge tokens together, but this *might not be necessary*, at least not for use cases we're familiar with.
+
+![Committer Flow](./committer_flow.png)
+
+#### Writer flow
+
+1. Read the branch's existing staging token: currently, we batch requests for a short duration and amortize the cost across requests.
```

Comment:
As far as I understand, this flow can cause a single write to appear in more than one commit. In particular, it can't be described as an idempotent operation in (3) below. Not saying this is necessarily bad but I'm finding it hard to describe it in terms of a consistency model.

In the following example, write 001 appears on commit `a`, reverted by commit `b` and then appears again as a new write on commit `c`.

Write 001 start – to staging token A
Commit `a` for token A
Write 002 start – to staging token B
Commit `b` for token B
Write 002 finalize: staging token was changed – write again to token C
Write 001 finalize: staging token was changed – write again to token C
Commit `c` for token C

Reply:
Update: talked f2f with @arielshaqed, who explained that this "limitation" was discussed by the team. @ozkatz can you please add an explanation about it?

Reply:
#2480 (created after F2F'ing with @ozkatz ) is my attempt to describe what guarantees we _do_ give.

Reply:
Great minds *do* think alike I guess. Thanks @arielshaqed for the description (should we maybe add it to this PR?)!